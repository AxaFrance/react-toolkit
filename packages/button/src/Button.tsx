import {
  withClickId,
  WithClickIdProps,
  compose,
  identity,
} from '@axa-fr/react-toolkit-core';

import ButtonCore, { ButtonCoreProps } from './ButtonCore';

export type ButtonProps = WithClickIdProps<ButtonCoreProps, 'onClick'>;

const Button = compose(
  identity<ButtonCoreProps>(),
  withClickId({ event: ['onClick'] })
)(ButtonCore);

Button.displayName = 'Button';

export default Button;
