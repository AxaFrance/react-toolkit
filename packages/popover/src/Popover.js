import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Constants } from '@axa-fr/react-toolkit-core';
import PopoverBase from './PopoverBase';
import PopoverPlacements from './PopoverPlacements';
import PopoverModes from './PopoverModes';

const propTypes = {
  ...Constants.propTypes,
  title: PropTypes.string,
  children: PropTypes.any,
  placement: PropTypes.oneOf([
    PopoverPlacements.top,
    PopoverPlacements.bottom,
    PopoverPlacements.left,
    PopoverPlacements.right,
  ]),
  mode: PropTypes.oneOf([PopoverModes.over, PopoverModes.click]),
};
const defaultClassName = 'af-popover__container';
const defaultProps = {
  ...Constants.defaultProps,
  className: defaultClassName,
  placement: PopoverPlacements.top,
  mode: PopoverModes.click,
};

class PopoverClick extends Component {
  constructor(props) {
    super(props);
    this.toggle = this.toggle.bind(this);
    this.outsideTap = this.outsideTap.bind(this);
    this.state = {
      isOpen: false,
    };
  }

  toggle() {
    const { isOpen } = this.state;
    this.setState({
      isOpen: !isOpen,
    });
  }

  outsideTap() {
    this.setState({
      isOpen: false,
    });
  }

  render() {
    const { children, placement, className, classModifier } = this.props;
    const { isOpen } = this.state;
    return (
      <PopoverBase
        className={className}
        classModifier={classModifier}
        onToggle={this.toggle}
        onOutsideTap={this.outsideTap}
        isOpen={isOpen}
        placement={placement}>
        {children}
      </PopoverBase>
    );
  }
}

class PopoverOver extends Component {
  constructor(props) {
    super(props);
    this.enter = this.enter.bind(this);
    this.leave = this.leave.bind(this);
    this.state = {
      isOpen: false,
    };
  }

  enter() {
    this.setState({
      isOpen: true,
    });
  }

  leave() {
    this.setState({
      isOpen: false,
    });
  }

  render() {
    const { children, placement, className, classModifier } = this.props;
    const { isOpen } = this.state;
    return (
      <span onMouseEnter={this.enter} onMouseLeave={this.leave}>
        <PopoverBase
          isOpen={isOpen}
          placement={placement}
          className={className}
          classModifier={classModifier}>
          {children}
        </PopoverBase>
      </span>
    );
  }
}

const Popover = props => {
  const { children, placement, className, classModifier, mode } = props;
  const DynamicComponent =
    mode === PopoverModes.click ? PopoverClick : PopoverOver;

  return (
    <DynamicComponent
      className={className}
      classModifier={classModifier}
      placement={placement}>
      {children}
    </DynamicComponent>
  );
};

Popover.Pop = PopoverBase.Pop;
Popover.Over = PopoverBase.Over;

Popover.propTypes = propTypes;
Popover.defaultProps = defaultProps;

export default Popover;
