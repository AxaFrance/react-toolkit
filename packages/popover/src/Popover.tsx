import React from 'react';
import { Constants } from '@axa-fr/react-toolkit-core';
import PopoverBase from './PopoverBase';
import PopoverModes from './PopoverModes';
import Placement from './PopoverPlacements';

const defaultProps = {
  ...Constants.defaultProps,
  className: 'af-popover__container',
  placement: 'top' as Placement,
  mode: PopoverModes.click,
};

type Props = Partial<typeof defaultProps> & {
  children?: React.ReactNode;
};

const PopoverClick = ({ children, placement, className, classModifier }: Props) => {
  const wrapperRef = React.useRef(null);
  const [isOpen, setOpen] = React.useState(false);
  const [isHover, setHover] = React.useState(false);
  const [isPopoverHover, setPopoverHover] = React.useState(false);

  const handleClickOutside = (event: MouseEvent | React.MouseEvent) => {
    if (wrapperRef.current && !wrapperRef.current.contains(event.target)) {
      setOpen(false);
      document.removeEventListener('click', handleClickOutside);
    }
  };

  const handleClick = (
    event: MouseEvent | React.MouseEvent | React.KeyboardEvent<HTMLDivElement>
  ) => {
    if (isPopoverHover) {
      event.stopPropagation();
      return;
    }

    const shouldOpen = !isOpen && isHover;
    setOpen(shouldOpen);

    if (shouldOpen) {
      document.addEventListener('click', handleClickOutside);
    }
  };

  const handleMouseEnter = () => {
    setHover(true);
  };

  const handleMouseLeave = () => {
    setHover(false);
  };

  const handleMouseEnterPopover = () => {
    setPopoverHover(true);
  };

  const handleMouseLeavePopover = () => {
    setPopoverHover(false);
  };

  return (
    <div
      role="button"
      tabIndex={0}
      ref={wrapperRef}
      className="af-popover__wrapper"
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      onKeyDown={handleClick}
      onClick={handleClick}
    >
      <PopoverBase
        onMouseEnter={handleMouseEnterPopover}
        onMouseLeave={handleMouseLeavePopover}
        isOpen={isOpen}
        placement={placement}
        className={className}
        classModifier={classModifier}
      >
        {children}
      </PopoverBase>
    </div>
  );
};

const PopoverOver = ({ children, placement, className, classModifier }: Props) => {
  const [isOpen, setOpen] = React.useState(false);

  const handleMouseEnter = () => {
    setOpen(true);
  };

  const handleMouseLeave = () => {
    setOpen(false);
  };

  return (
    <div
      className="af-popover__wrapper"
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <PopoverBase
        isOpen={isOpen}
        placement={placement}
        className={className}
        classModifier={classModifier}
      >
        {children}
      </PopoverBase>
    </div>
  );
};

export { PopoverClick, PopoverOver };

const Popover = ({ children, placement, className, classModifier, mode }: Props) => {
  return mode === PopoverModes.click ? (
    <PopoverClick className={className} classModifier={classModifier} placement={placement}>
      {children}
    </PopoverClick>
  ) : (
    <PopoverOver className={className} classModifier={classModifier} placement={placement}>
      {children}
    </PopoverOver>
  );
};

Popover.Pop = PopoverBase.Pop;
Popover.Pop.displayName = 'Popover.Pop';

Popover.Over = PopoverBase.Over;
Popover.Over.displayName = 'Popover.Over';

Popover.defaultProps = defaultProps;

export default Popover;
