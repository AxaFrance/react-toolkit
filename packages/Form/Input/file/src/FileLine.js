import React from 'react';
import PropTypes from 'prop-types';
import { ClassManager } from '@axa-fr/react-toolkit-core';
import Button from '@axa-fr/react-toolkit-button';
import Popover from '@axa-fr/react-toolkit-popover';

const propTypes = {
  file: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  className: PropTypes.string,
  classModifier: PropTypes.string,
};
const defaultClassName = 'af-form__file-line';
const defaultProps = {
  disabled: false,
  className: defaultClassName,
  classModifier: null,
};

const style = {
  maxWidth: '200px',
  maxHeight: '200px',
};

export const Preview = ({ file }) => {
  if (file && file.type && file.type.startsWith('image')) {
    return (
      <Popover mode="hover">
        <Popover.Pop>
          <img
            src={file.preview}
            style={style}
            className="af-form__file-thumbnail"
            alt="File Preview"
          />
        </Popover.Pop>
        <Popover.Over>
          <i className="glyphicon glyphicon-picture" />
        </Popover.Over>
      </Popover>
    );
  }
  return <i className="glyphicon glyphicon-file" />;
};

const FileLine = ({
  className,
  classModifier,
  file,
  disabled,
  id,
  onClick,
}) => {
  const componentClassName = ClassManager.getComponentClassName(
    className,
    classModifier,
    defaultClassName
  );
  return (
    <li className={componentClassName}>
      <Preview file={file} />
      <span>{file.name}</span>
      <span>{file.size}</span>

      <Button
        disabled={disabled}
        type="button"
        className="af-link"
        classModifier="delete-file"
        onClick={() => onClick(id)}>
        <span className="af-link__text">Supprimer</span>
      </Button>
    </li>
  );
};

FileLine.propTypes = propTypes;
FileLine.defaultProps = defaultProps;

export default FileLine;
