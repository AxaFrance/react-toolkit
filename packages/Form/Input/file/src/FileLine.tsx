import React from 'react';
import { ClassManager } from '@axa-fr/react-toolkit-core';
import Button from '@axa-fr/react-toolkit-button';
import Popover from '@axa-fr/react-toolkit-popover';

const defaultClassName = 'af-form__file-line';
const style = {
  maxWidth: '200px',
  maxHeight: '200px',
};

type PreviewProps = Pick<Props, 'file'>;

export const Preview = ({ file }: PreviewProps) => {
  if (file && file.type && file.type.startsWith('image')) {
    return (
      <Popover mode="hover">
        <Popover.Pop>
          <img
            src={file.preview}
            style={style}
            className="af-form__file-thumbnail"
            alt="File Preview"
          />
        </Popover.Pop>
        <Popover.Over>
          <i className="glyphicon glyphicon-picture" />
        </Popover.Over>
      </Popover>
    );
  }
  return <i className="glyphicon glyphicon-file" />;
};

type File = { type?: string; preview: string; name: string; size: number };
type Props = {
  file: File;
  id: string;
  onClick?: (id: string) => void;
  className?: string;
  classModifier?: string;
  disabled?: boolean;
};
const FileLine = ({
  className,
  classModifier,
  file,
  disabled,
  id,
  onClick,
}: Props) => {
  const componentClassName = ClassManager.getComponentClassName(
    className,
    classModifier,
    defaultClassName
  );
  return (
    <li className={componentClassName}>
      <Preview file={file} />
      <span>{file.name}</span>
      <span>{file.size}</span>

      <Button
        disabled={disabled}
        type="button"
        className="af-link"
        classModifier="delete-file"
        onClick={() => onClick(id)}>
        <span className="af-link__text">Supprimer</span>
      </Button>
    </li>
  );
};

export default FileLine;
