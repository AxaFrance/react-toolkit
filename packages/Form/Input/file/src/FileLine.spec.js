import React from 'react';
import { shallow } from 'enzyme';

import Button from '@axa-fr/react-toolkit-button';
import Popover from '@axa-fr/react-toolkit-popover';
import FileLine, { Preview } from './FileLine';

describe('<File.FileInput>', () => {
  it('renders File.FileInput correctly', () => {
    const wrapper = shallow(
      <FileLine file={{ name: 'name', size: 1, preview: '#' }} id="id" />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('Should called onClick function when button have been clicked', () => {
    const onClickMock = jest.fn();

    const wrapper = shallow(
      <FileLine
        file={{ name: 'name', size: 1, preview: '#' }}
        id="id"
        onClick={onClickMock}
      />
    );
    wrapper.find(Button).simulate('click');
    expect(onClickMock).toBeCalled();
  });

  it('renders Preview correctly for type image', () => {
    const file = { type: 'image' };
    const wrapper = shallow(<Preview file={file} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders Preview correctly for other type', () => {
    const file = { type: 'pdf' };
    const wrapper = shallow(<Preview file={file} />);
    expect(wrapper).toMatchSnapshot();
  });
});
