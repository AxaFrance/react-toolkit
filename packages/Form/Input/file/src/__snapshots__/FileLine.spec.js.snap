// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<File.FileInput> renders File.FileInput correctly 1`] = `
<li
  className="af-form__file-line"
>
  <Preview
    file={
      Object {
        "name": "name",
        "preview": "#",
        "size": 1,
      }
    }
  />
  <span>
    name
  </span>
  <span>
    1
  </span>
  <Button
    classModifier="delete-file"
    className="af-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      className="af-link__text"
    >
      Supprimer
    </span>
  </Button>
</li>
`;

exports[`<File.FileInput> renders Preview correctly for other type 1`] = `
<i
  className="glyphicon glyphicon-file"
/>
`;

exports[`<File.FileInput> renders Preview correctly for type image 1`] = `
<Popover
  classModifier={null}
  className="af-popover__container"
  mode="hover"
  placement="top"
>
  <Popover.Pop>
    <img
      alt="File Preview"
      className="af-form__file-thumbnail"
      style={
        Object {
          "maxHeight": "200px",
          "maxWidth": "200px",
        }
      }
    />
  </Popover.Pop>
  <Popover.Over>
    <i
      className="glyphicon glyphicon-picture"
    />
  </Popover.Over>
</Popover>
`;
